jupyterkernelgen/jupyterkernelgen.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+import os
+import glob
+import shutil
+import sys
+import readline
+import subprocess
+
+# allow tab completion in input()
+readline.set_completer_delims(' \t\n=')
+readline.parse_and_bind("tab: complete")
+
+# define ANSI escape codes for coloured text
+class colors:
+    HEADER = '\033[95m'
+    OK = '\033[92m'
+    WARNING = '\033[93m'
+    FAIL = '\033[91m'
+    ENDC = '\033[0m'
+
+def check_for_conda():
+    print(f"{colors.HEADER}LOOKING FOR CONDA...{colors.ENDC}")
+
+    conda_exe = None
+    try:
+        conda_exe = shutil.which("mamba") # first try getting mamba because its faster
+    except shutil.Error as e:
+        print(f"{colors.FAIL}error occurred getting mamba from path: {e}{colors.ENDC}", file=sys.stderr)
+
+    if conda_exe == None:
+        try:
+            conda_exe = shutil.which("conda") # try finding conda if mamba isn't available
+        except shutil.Error as e:
+            print(f"{colors.FAIL}error occurred getting conda from path: {e}{colors.ENDC}", file=sys.stderr)
+            exit(1)
+
+        if conda_exe == None:
+            print(f"{colors.FAIL}no conda executable on the path. Exiting...{colors.ENDC}", file=sys.stderr)
+            sys.exit(1)
+
+    print(f"- found conda executable: {colors.OK}{conda_exe}{colors.ENDC}\n")
+
+    return conda_exe
+
+def get_conda_env():
+    try:
+        # replace ~ with user's home directory and make it an absolute path
+        conda_env = os.path.abspath(os.path.expanduser(input(f"{colors.HEADER}ENTER PATH TO CONDA ENVIRONMENT{colors.ENDC}: ")))
+
+        if os.path.isdir(conda_env + "/conda-meta"): # all conda environments have a /conda-meta directory
+            print(f"- found conda env: {colors.OK}{conda_env}{colors.ENDC}\n")
+        else:
+            print(f"{colors.FAIL}given conda env path is not a conda environment. Exiting...{colors.ENDC}", file=sys.stderr)
+            sys.exit(1)
+
+    except OSError as e:
+        print(f"{colors.FAIL}error occurred checking directory: {e}{colors.ENDC}", file=sys.stderr)
+        sys.exit(1)
+
+    return conda_env
+
+def ipykernel_installed(conda_env):
+    print(f"{colors.HEADER}CHECKING FOR IPYTHON IN CONDA ENV...{colors.ENDC}")
+
+    # We check for both ipykernel and ipython to be installed just to be doubly sure
+    # that we have everything we need
+    found_kernel = False
+    for file in glob.glob(conda_env + "/lib/python3.*/site-packages/ipykernel"):
```

Comment:
Perhaps another approach could be to activate the conda environment and see if you can run `ipython --version` . What do you think?

Reply:
yeah I like that. It does seem a little more stable than checking for files.

Reply:
Having tried that, it seems a little trickier than initially thought. Even when I try to source the mamba initialization script, python doesn't recognize ipython even if it is installed because it doesn't actually use the shell. Determining if there is a file at `{conda_env}/bin/ipython` effectively does the same thing without having to deal with sourcing any scripts.